tools/validate_spec.py

```diff
@@ -161,6 +215,112 @@ def _validate_message(
+
+        self.total_links = sum(
+            len(structure) for structure in self.__total_message_coverage.values()
+        )
+        self.total_covered_links = 0
+
+    def update(self, message_value: MessageValue) -> None:
+        if self.__coverage:
+            messages = message_value.inner_messages() + [message_value]
+            for message in messages:
+                for link in message.path:
+                    if not self.__total_message_coverage[message.identifier][link]:
+                        self.total_covered_links += 1
+                        self.__total_message_coverage[message.identifier][link] = True
+
+    def file_total_links(self, file_name: str) -> int:
+        assert file_name in self.__spec_files
+        return sum(
+            len(self.__total_message_coverage[message]) for message in self.__spec_files[file_name]
+        )
+
+    def file_covered_links(self, file_name: str) -> int:
+        assert file_name in self.__spec_files
+        return sum(
+            list(self.__total_message_coverage[message].values()).count(True)
+            for message in self.__spec_files[file_name]
+        )
+
+    def file_uncovered_links(self, file_name: str) -> List[Link]:
+        assert file_name in self.__spec_files
+        return list(
+            itertools.chain(
+                *[
+                    [
+                        link
+                        for link, covered in self.__total_message_coverage[message].items()
+                        if not covered
+                    ]
+                    for message in self.__spec_files[file_name]
+                ]
+            )
+        )
+
+    def print_coverage(self) -> None:
+        if self.__coverage:
+            self.__print_coverage_overview()
+            if self.total_covered_links / self.total_links != 1:
+                self.__print_link_coverage()
+
+    def __print_coverage_overview(self) -> None:
+        print("\n")
+        print("-" * 80)
+        print(f"{'RecordFlux Validation Coverage Report' : ^80}")
+        print(f"Directory: {os.getcwd()}")
+        print("-" * 80)
+        print(f"{'File' : <40} {'Links' : >10} {'Used' : >10} {'Coverage' : >15}")
+        for file in self.__spec_files:
+            file_links = self.file_total_links(file)
+            file_covered_links = self.file_covered_links(file)
+            print(
+                f"{file : <40} {file_links : >10} {file_covered_links : >10} "
+                f"{file_covered_links / file_links :>15.2%}"
+            )
+        print("-" * 80)
+        print(
+            f"{'TOTAL' : <40} {self.total_links: >10} {self.total_covered_links : >10} "
+            f"{self.total_covered_links / self.total_links :15.2%}"
+        )
+        print("-" * 80)
+
+    def __print_link_coverage(self) -> None:
+        print("\n")
+        print("=" * 80)
+        print(f"{'Missing Link Coverage' : ^80}")
+        print("=" * 80)
```

Comment:
What is a "missing link coverage"? Did you mean "uncovered links"?

Reply:
I did not like "Uncovered Links" because it sounds like someone "discovered" the links by accident. Probably that's just me. Changed it in 19aca4c5.

Reply:
Actually I never thought about the different meaning of "uncovered" in this context. AFAICT "uncovered lines" is a usual term wrt. to test coverage, so I think it makes sense to use a similar term here.